### PR TITLE
Fixes #28246 - Create bulk cancel api

### PR DESCRIPTION
**Pull request by MariaAga**
Adds a way to bulk cancel tasks using a query or an array of ids.
Return information about canceled tasks and failed to cancel tasks

### CODE REVIEW

File: app/controllers/foreman_tasks/api/tasks_controller.rb [hunk 1]
```diff
@@ -122,6 +122,42 @@ def bulk_resume
...
 
+      api :POST, '/tasks/bulk_cancel', N_('Cancel all cancellable tasks')
+      param :search, String, :desc => N_('Cancel tasks matching search string')
+      param :task_ids, Array, :desc => N_('Cancel specific tasks by ID')
+      def bulk_cancel
+        scope = resource_scope
+        scope = scope.search_for(params[:search]) if params[:search]
+        scope = scope.select('DISTINCT foreman_tasks_tasks.*')
+        if params[:search].nil? && params[:task_ids].nil?
+          scope = scope.where(:state => :scheduled)
```
Comment (MariaAga) [inline app/controllers/foreman_tasks/api/tasks_controller.rb]: @adamruzicka Is this correct? I'm not sure if there are other kind of tasks that can be canceled

File: app/controllers/foreman_tasks/api/tasks_controller.rb [hunk 2]
```diff
@@ -122,6 +121,42 @@ def bulk_resume
...
+      api :POST, '/tasks/bulk_cancel', N_('Cancel all cancellable tasks')
+      param :search, String, :desc => N_('Cancel tasks matching search string')
+      param :task_ids, Array, :desc => N_('Cancel specific tasks by ID')
+      def bulk_cancel
+        scope = resource_scope
+        scope = scope.search_for(params[:search]) if params[:search]
+        scope = scope.select('DISTINCT foreman_tasks_tasks.*')
+        if params[:search].nil? && params[:task_ids].nil?
+          raise ArgumentError, _('Please provide at least one of search or task_ids parameters in the request')
+        end
```
Comment (adamruzicka) [inline app/controllers/foreman_tasks/api/tasks_controller.rb]: I would move this to the top of the method

File: app/controllers/foreman_tasks/api/tasks_controller.rb [hunk 3]
```diff
@@ -122,6 +121,42 @@ def bulk_resume
...
+            begin
+              task.cancel
+              cancelled << task_hash(task)
+            rescue RuntimeError
+              failed << task_hash(task)
+            end
+          else
+            skipped << task_hash(task)
+          end
+        end
```
Comment (adamruzicka) [inline app/controllers/foreman_tasks/api/tasks_controller.rb]: How about using `Enumerable#partition` to first split the scope into tasks which can be cancelled and which cannot and then operating only on those which can be cancelled?

File: app/controllers/foreman_tasks/api/tasks_controller.rb [hunk 4]
```diff
@@ -122,6 +121,42 @@ def bulk_resume
...
+
+        cancelled = []
+        failed = []
+        skipped = []
+        scope.each do |task|
+          if task.cancellable?
+            begin
+              task.cancel
+              cancelled << task_hash(task)
+            rescue RuntimeError
```
Comment (adamruzicka) [inline app/controllers/foreman_tasks/api/tasks_controller.rb]: An exception should be raised only if the task's execution plan cannot be loaded. However all tasks which are `cancellable?` should have an execution plan. That would mean we can drop the `begin-rescue-end` block

File: app/models/foreman_tasks/task.rb [hunk 5]
```diff
@@ -123,6 +123,10 @@ def pending?
     def resumable?
       false
     end
+    
+    def cancellable?
+      execution_plan&.cancellable?
+    end
```
Comment (adamruzicka) [inline app/models/foreman_tasks/task.rb]: This is already defined on `ForemanTasks::Task::DynflowTask` and should not be needed.

File: app/controllers/foreman_tasks/api/tasks_controller.rb (hunk 4 above)
Comment (MariaAga) [reply app/controllers/foreman_tasks/api/tasks_controller.rb]: Does this mean task canceling can't fail?

File: app/models/foreman_tasks/task.rb (hunk 5 above)
Comment (MariaAga) [reply app/models/foreman_tasks/task.rb]: Removed, thanks!

File: app/controllers/foreman_tasks/api/tasks_controller.rb (hunk 3 above)
Comment (MariaAga) [reply app/controllers/foreman_tasks/api/tasks_controller.rb]: Looks much better that way, thanks!

File: app/controllers/foreman_tasks/api/tasks_controller.rb (hunk 2 above)
Comment (MariaAga) [reply app/controllers/foreman_tasks/api/tasks_controller.rb]: Moved this and the second one up

File: app/controllers/foreman_tasks/api/tasks_controller.rb (hunk 4 above)
Comment (adamruzicka) [reply app/controllers/foreman_tasks/api/tasks_controller.rb]: If a task is `cancellable?` then the cancelling should not raise, unless something happened between calls to `cancellable?` and `cancel`.

Under normal conditions, if a task is `cancellable?` then the worst thing that can happen is that the cancel won't actually do anything.

File: app/controllers/foreman_tasks/api/tasks_controller.rb [hunk 6]
```diff
@@ -311,6 +323,13 @@ def tasks_list(filtered_scope)
           results: results
         }
       end
+
+      def bulk_scope
+        scope = resource_scope
+        scope = scope.search_for(params[:search]) if params[:search]
+        scope = scope.select('DISTINCT foreman_tasks_tasks.*')
+        scope.where(:id => params[:task_ids]) if params[:task_ids]
```
[the remark this answers is not in the document]
Comment (MariaAga) [reply app/controllers/foreman_tasks/api/tasks_controller.rb]: added, thanks!

File: app/controllers/foreman_tasks/api/tasks_controller.rb [hunk 7]
```diff
@@ -89,19 +89,13 @@ def bulk_search
       param :search, String, :desc => N_('Resume tasks matching search string')
       param :task_ids, Array, :desc => N_('Resume specific tasks by ID')
       def bulk_resume
-        scope = resource_scope
-        scope = scope.search_for(params[:search]) if params[:search]
-        scope = scope.select('DISTINCT foreman_tasks_tasks.*')
         if params[:search].nil? && params[:task_ids].nil?
-          scope = scope.where(:state => :paused)
-          scope = scope.where(:result => :error)
+          raise ArgumentError, _('Please provide at least one of search or task_ids parameters in the request')
```
Comment (adamruzicka) [inline app/controllers/foreman_tasks/api/tasks_controller.rb]: See https://github.com/theforeman/foreman-tasks/pull/517/files#r394391514

Comment (MariaAga) [reply app/controllers/foreman_tasks/api/tasks_controller.rb]: Changed to `BadRequest`, Thanks!

**Comment by adamruzicka** [comment]
Thank you @MariaAga !

**Pull request merged**